[SPARK-34432][SQL][TESTS] Add JavaSimpleWritableDataSource

kevincmchen opened the pull request:
### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/19269

In #19269 , there is only a scala implementation of simple writable data source in `DataSourceV2Suite`.

This PR adds a java implementation of it.

### Why are the changes needed?

To improve test coverage.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing testsuites

kevincmchen committed:
[SPARK-34432][SQL][TEST][FOLLOWUP] Add a java implementation of simple writable data source in `DataSourceV2Suite`

### What changes were proposed in this pull request?

This is a followup of https://github.com/apache/spark/pull/19269

In #19269 , there is only a scala implementation of simple writable data source in `DataSourceV2Suite`.

This PR adds a java implementation of it.

### Why are the changes needed?

To improve test coverage.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing testsuites

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
cc @rdblue @cloud-fan

dongjoon-hyun commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Arrays;
+import java.util.Iterator;
```
nit. Could you re-order the import like the following grouping?
```java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Iterator;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.*;

import org.apache.spark.deploy.SparkHadoopUtil;
import org.apache.spark.sql.catalyst.InternalRow;
import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
import org.apache.spark.sql.connector.TestingV2Source;
import org.apache.spark.sql.connector.catalog.SessionConfigSupport;
import org.apache.spark.sql.connector.catalog.SupportsWrite;
import org.apache.spark.sql.connector.catalog.Table;
import org.apache.spark.sql.connector.read.InputPartition;
import org.apache.spark.sql.connector.read.PartitionReader;
import org.apache.spark.sql.connector.read.PartitionReaderFactory;
import org.apache.spark.sql.connector.read.ScanBuilder;
import org.apache.spark.sql.connector.write.*;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.sql.util.CaseInsensitiveStringMap;
import org.apache.spark.util.SerializableConfiguration;
```

kevincmchen replied:
I have already fixed it, thx a lot.

dongjoon-hyun commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+import java.util.Arrays;
+import java.util.Iterator;
+
+public class JavaSimpleWritableDataSource implements TestingV2Source, SessionConfigSupport {
```
It would be great if you add a class description here, @kevincmchen .

dongjoon-hyun replied:
Do we need to use `TestingV2Source`? Shall we use `SimpleTableProvider` instead?

cloud-fan replied:
Yea I think we should follow the scala version and use `SimpleTableProvider`. `TestingV2Source.inferSchema` doesn't match the schema here.

kevincmchen replied:
> Yea I think we should follow the scala version and use `SimpleTableProvider`. `TestingV2Source.inferSchema` doesn't match the schema here.

using SimpleTableProvider will cause  compile error and the error is 
`Error:(53, 8) java: test.org.apache.spark.sql.connector.JavaSimpleWritableDataSource is not abstract and does not override abstract method org$apache$spark$sql$internal$connector$SimpleTableProvider$$loadedTable_$eq(org.apache.spark.sql.connector.catalog.Table) in org.apache.spark.sql.internal.connector.SimpleTableProvider`

the reason is that scala traits cannot be extended in Java if they have any implementation details.

kevincmchen replied:
so i have 2 options:
1. create a java implementation of SimpleTableProvider
2. use TestingV2Source and override inferSchema function.

Do you think which one is better?

cloud-fan replied:
Ah, I see. Makes sense.

Maybe we can change the schema of this writable v2 source (as well as the read part), to match `TestingV2Source`, so that the scala version can also use `TestingV2Source`, to be consistent with the java version.

dongjoon-hyun commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+    public void close() {}
+  }
+
+  class MyScanBuilder extends JavaSimpleScanBuilder {
```
In general, please match the class layout with the existing `SimpleWritableDataSource`.
For example, please move this to the top as the first subclass of this.

kevincmchen replied:
I have already fixed it, thx a lot.

dongjoon-hyun commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+        } else {
+          return new InputPartition[0];
+        }
+      } catch (Exception e) {
```
This test case can be a reference model for the other developers. Are you suggesting `Java Data Source` should follow this kind of `Exception` wrapping?

cloud-fan replied:
We should at least catch the specific exception types.

kevincmchen replied:
I have already fixed it, thx a lot.

dongjoon-hyun commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+    }
+
+    @Override
+    public Write build() {
```
Thanks for using a new API instead of the deprecated ones.

dongjoon-hyun commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+    private final String queryId;
+    private final boolean needTruncate;
+
+    MyWrite(String path, String queryId, boolean needTruncate) {
```
`CaseInsensitiveStringMap` is a Java class. Can we use it instead of having this complex constructor?

dongjoon-hyun commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+    @Override
+    public StructType schema() {
+      return tableSchema;
+    }
```
Ur, we need `capabilities` because we have `SupportsWrite`.

cloud-fan replied:
This seems like an unnecessary inconsistency, as before this PR `JavaSimpleBatchTable` doesn't need to have write table capability at all.

We can probably remove `BATCH_WRITE` and `TRUNCATE` from `JavaSimpleBatchTable` and add it here, to follow the scala version.

kevincmchen replied:
I have already fixed it, thx a lot.

dongjoon-hyun reviewed (commented):
Thank you for making your first contribution. I left a few general comments.

cloud-fan commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,374 @@
+    }
+  }
+
+  static class JavaSimpleCounter {
```
The test checks `object SimpleCounter` and the java version should update it as well, instead of creating a new counter, otherwise the test fails.

kevincmchen replied:
ok, I will use `object SimpleCounter`

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #135199 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135199/testReport)** for PR 31560 at commit [`d5dc307`](https://github.com/apache/spark/commit/d5dc3075ecea59dfc5949b2bd5b400c51b9c37c8).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39780/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39780/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39780/

SparkQA commented:
**[Test build #135199 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135199/testReport)** for PR 31560 at commit [`d5dc307`](https://github.com/apache/spark/commit/d5dc3075ecea59dfc5949b2bd5b400c51b9c37c8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaSimpleWritableDataSource implements TestingV2Source, SessionConfigSupport `
  * `  static class JavaCSVInputPartitionReader implements InputPartition `
  * `  static class JavaCSVReaderFactory implements PartitionReaderFactory `
  * `  static class JavaSimpleCounter `
  * `  static class JavaCSVDataWriterFactory implements DataWriterFactory `
  * `  static class JavaCSVDataWriter implements DataWriter<InternalRow> `
  * `  class MyScanBuilder extends JavaSimpleScanBuilder `
  * `  static class MyWriteBuilder implements WriteBuilder, SupportsTruncate `
  * `  static class MyWrite implements Write `
  * `  static class MyBatchWrite implements BatchWrite `
  * `  class MyTable extends JavaSimpleBatchTable implements SupportsWrite `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135199/

kevincmchen committed:
ADD:
	1. add a class description for JavaSimpleWritableDataSource

OPTIMIZE:
	1. re-order the import
	2. match the class layout with the existing SimpleWritableDataSource

UPDATE:
    1. catch the specific exception(IOException) instead of Exception.
    2. use SimpleCounter

SparkQA commented:
**[Test build #135230 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135230/testReport)** for PR 31560 at commit [`88c6edf`](https://github.com/apache/spark/commit/88c6edf49ac19549aba9aa5ae04e6b7c5665191f).

kevincmchen committed:
Update JavaSimpleWritableDataSource.java

delete duplicated blank Lines in `JavaSimpleWritableDataSource`

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39811/

SparkQA commented:
**[Test build #135231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135231/testReport)** for PR 31560 at commit [`8c080ba`](https://github.com/apache/spark/commit/8c080ba80ced4af173b9d2cebc9475e0a722c75e).

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39811/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39811/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39812/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39812/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39812/

SparkQA commented:
**[Test build #135230 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135230/testReport)** for PR 31560 at commit [`88c6edf`](https://github.com/apache/spark/commit/88c6edf49ac19549aba9aa5ae04e6b7c5665191f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  static class JavaCSVInputPartitionReader implements InputPartition `
  * `  static class JavaCSVReaderFactory implements PartitionReaderFactory `
  * `  static class JavaCSVDataWriterFactory implements DataWriterFactory `
  * `  static class JavaCSVDataWriter implements DataWriter<InternalRow> `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135230/

SparkQA commented:
**[Test build #135231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135231/testReport)** for PR 31560 at commit [`8c080ba`](https://github.com/apache/spark/commit/8c080ba80ced4af173b9d2cebc9475e0a722c75e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135231/

cloud-fan commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,370 @@
+          return new InputPartition[0];
+        }
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```
This seems not java friendly, I think we should update `Batch#planInputPartitions` to add `throws IOException`, the same as `PartitionReader#next`.

The `throws` clause is a compile-time check and won't break binary compatibility. Also cc @rdblue

kevincmchen replied:
I have some different opinions recently. if we throw IOException in Batch#planInputPartitions

cloud-fan commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,370 @@
+        if (fs.exists(dataPath)) {
+          return Arrays.stream(fs.listStatus(dataPath))
+                  .filter(
+                          status -> {
```
nit: 2 space indentation

kevincmchen committed:
change the schema of `JavaSimpleWritableDataSource` to match `TestingV2Source`

kevincmchen committed:
fix conflicts

kevincmchen committed:
reformat code

SparkQA commented:
**[Test build #135307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135307/testReport)** for PR 31560 at commit [`9037d5f`](https://github.com/apache/spark/commit/9037d5f72161fd13c9d197842bfe29ab4ee48cb3).

SparkQA commented:
**[Test build #135310 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135310/testReport)** for PR 31560 at commit [`9178240`](https://github.com/apache/spark/commit/9178240ee37b9b4abb318b91f6ce2bbf4c0c6f84).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39887/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39890/

kevincmchen committed:
In order to match SimpleBatchTable, remove BATCH_WRITE and TRUNCATE from JavaSimpleBatchTable and override the function capabilities in MyTable

SparkQA commented:
**[Test build #135313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135313/testReport)** for PR 31560 at commit [`3fc7288`](https://github.com/apache/spark/commit/3fc7288e1ff49510174545aabbcc028b409cec5c).

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39890/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39893/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39890/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39893/

SparkQA commented:
**[Test build #135307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135307/testReport)** for PR 31560 at commit [`9037d5f`](https://github.com/apache/spark/commit/9037d5f72161fd13c9d197842bfe29ab4ee48cb3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135307/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39893/

SparkQA commented:
**[Test build #135310 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135310/testReport)** for PR 31560 at commit [`9178240`](https://github.com/apache/spark/commit/9178240ee37b9b4abb318b91f6ce2bbf4c0c6f84).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135310/

SparkQA commented:
**[Test build #135313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135313/testReport)** for PR 31560 at commit [`3fc7288`](https://github.com/apache/spark/commit/3fc7288e1ff49510174545aabbcc028b409cec5c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135313/

kevincmchen committed:
reformat code

SparkQA commented:
**[Test build #135320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135320/testReport)** for PR 31560 at commit [`c665f68`](https://github.com/apache/spark/commit/c665f683f708a328cc4afdfec57e6e54d04949a7).

kevincmchen committed:
reformat code

SparkQA commented:
**[Test build #135322 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135322/testReport)** for PR 31560 at commit [`0d85f08`](https://github.com/apache/spark/commit/0d85f08a9e41322e62398ab45ba78046f8d2a78c).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39900/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39902/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39902/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39902/

SparkQA commented:
**[Test build #135320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135320/testReport)** for PR 31560 at commit [`c665f68`](https://github.com/apache/spark/commit/c665f683f708a328cc4afdfec57e6e54d04949a7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135320/

SparkQA commented:
**[Test build #135322 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135322/testReport)** for PR 31560 at commit [`0d85f08`](https://github.com/apache/spark/commit/0d85f08a9e41322e62398ab45ba78046f8d2a78c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135322/

cloud-fan commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,371 @@
+
+  @Override
+  public String keyPrefix() {
+    return "javaSimpleWritableDataSource";
```
Seems like we don't need to implement `SessionConfigSupport`. Can you open a follow-up PR to remove it from both the Scala and Java versions?

kevincmchen replied:
all right

cloud-fan commented on sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java:
```diff
@@ -0,0 +1,371 @@
+            Object[] objects =
+                Arrays.stream(currentLine.split(","))
+                    .map(String::trim)
+                    .map(Integer::parseInt)
```
can we change the Scala version as well in a follow-up PR?

kevincmchen replied:
ok，i'll change the scala version

cloud-fan commented:
thanks, merging to master!

cloud-fan commented:
@kevincmchen can you open a new PR to add `throws` clause to `planInputPartitions`, `commit`, etc., and trigger the discussion there?

kevincmchen commented:
> @kevincmchen can you open a new PR to add `throws` clause to `planInputPartitions`, `commit`, etc., and trigger the discussion there?

ok, all right.
